Add scene support to roborock

regevbr opened the pull request:
## Proposed change
Add ability to start scenes as added in https://github.com/Python-roborock/python-roborock/pull/317

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/37286
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

regevbr committed:
feature: add scene buttons to roborock

regevbr commented:
@Lash-L  this is still pending the release of the feature, and not yet tested, but I would appreciate your initial review

allenporter commented:
I don't think buttons are the right way to surface scenes.  Are scene entities a fit for this?

regevbr commented:
> I don't think buttons are the right way to surface scenes. Are scene entities a fit for this?

No, because ha scenes are for reproducing states of several entities. Here the name is misleading, the app calls it routine or program and it just fires a custom predefined program of the vacuum

allenporter commented:
> > I don't think buttons are the right way to surface scenes. Are scene entities a fit for this?
> 
> No, because ha scenes are for reproducing states of several entities. Here the name is misleading, the app calls it routine or program and it just fires a custom predefined program of the vacuum

Sticking to the first part: I don't think buttons are the right way to surface this.

regevbr commented:
Changed to scene.  
Once released I will create a separate initial PR to update the version of the lib first

regevbr commented:
upgrade pr - https://github.com/home-assistant/core/pull/137244

regevbr commented:
Validated locally that it works :-)
![image](https://github.com/user-attachments/assets/f228c5c8-32f5-4638-a7e3-af0c44b4cc70)

regevbr commented:
@Lash-L ready for review

regevbr committed:
feature: upgrade python-roborock

regevbr commented:
Added tests to fully cover new logic and created PR to update the docs

allenporter commented on homeassistant/components/roborock/rest.py:
```diff
@@ -0,0 +1,54 @@
+        try:
+            return await self._api_client.get_scenes(self._user_data, self._device.duid)
+        except RoborockException as err:
+            _LOGGER.warning(
```
This does not seem like appropriate error handling to just log warnings and return an empty result like nothing happened.

allenporter replied:
If we're doing this because not all devices support scenes then we should probably figure out which devices support them?

Lash-L replied:
So doing a test on my side - my device does not support scenes, and it instead just gives an empty list.

So RoborockException would not be valid as a catch for not supported devices and definitely shouldn't be ignored as it is now.

So I think calling get scenes and then setting up the entities will work - as we will get a empty list for devices that don't support it and therefore setup no entities. Figuring out explicitly what devices support scenes will likely be a harder task, and I think since it just returns an empty list, I would think it would be fine just to not setup any entities.

regevbr replied:
Ok I will remove. The thought behind it was to not cause the device to fail to setup just because of scenes but now that the logic to handle it is contained in the entity (and not init) it doesn't really matter.

allenporter replied:
Yep, understood the thought and intention.  

If there was a transient error that was just swallowed and we continue anyway, then the user would just randomly miss entities which isn't desirable.

There are specific best practices for retrying errors with backoff, etc, and incidicating status about the integration or entity, and error message to the user so raising HomeAssistantError is the way to go here.   If its a non-retryable error (e.g. bad request) then that root cause would need to be fixed to not happen in the first place.

allenporter commented on homeassistant/components/roborock/rest.py:
```diff
@@ -0,0 +1,54 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class RoborockRestApi:
```
I don't think this class is needed. It looks like the API for `RoborockApiClient` is incorrect and it should take these as constructor arguments like the other clients do.

regevbr replied:
Can you please further explain what you mean and what you need fixing?  
I guess you are talking about the fact that a function requires to get the user data as param instead of having it in the constructor?  
Even if that was the case, we would still want to wrap the functionality in our own class.

regevbr replied:
@Lash-L thoughts?

allenporter replied:
> Can you please further explain what you mean and what you need fixing? 

I'm saying make it look how the other APIs look, rather than a new style.

I'm saying (1) remove this class (2) update the class in the upstream library to take these constructors arguments.

> I guess you are talking about the fact that a function requires to get the user data as param instead of having it in the constructor? 

That's part of it.

> Even if that was the case, we would still want to wrap the functionality in our own class.

Why? The other APIs don't have this and call the api directly. I don't see why this is special.

Lash-L replied:
> update the class in the upstream library to take these constructors arguments

The RoborockApiClient should not take these arguments. As the RoborockApiClient is used for finding out what devices exist in the first place, whereas these parameters imply that you already know what devices you have.

Perhaps the coordinator could just have the api included like is currently done and these functions can live on the coordinator? Not sure - but I also don't think a RoborockRestAPI class needs to exist.

allenporter replied:
Sounds good to me.

regevbr replied:
Done

allenporter commented on homeassistant/components/roborock/scene.py:
```diff
@@ -0,0 +1,86 @@
+) -> None:
+    """Set up scene platform."""
+    coordinator_entities = await asyncio.gather(
+        *_build_setup_functions(config_entry.runtime_data.values()),
```
This 3 layers of nesting of function calls seems like it should just all be inlined here.

regevbr replied:
Did my best to inline them, but Im not a strong Python coder and getting an inline coroutine was a failed attempt on my side, Please let me know how I can do it even better if possible

Lash-L commented on homeassistant/components/roborock/rest.py:
```diff
@@ -0,0 +1,54 @@
+            _LOGGER.error("Failed to execute scene %s %s", scene_id, err)
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="command_failed",
```
you have to add to strings.json for these errors

Lash-L commented on homeassistant/components/roborock/coordinator.py:
```diff
@@ -168,19 +200,9 @@ async def set_current_map_rooms(self) -> None:
-        return slugify(self.duid)
-
 
 class RoborockDataUpdateCoordinatorA01(
```
A01 devices cannot do scenes, so this refactor is not needed

regevbr replied:
done

Lash-L commented on homeassistant/components/roborock/scene.py:
```diff
@@ -0,0 +1,86 @@
+    ) -> None:
+        """Create a scene entity."""
+        super().__init__(
+            f"{scene.name}_{coordinator.duid_slug}",
```
Don't scenes have a id associated with them? Let's do that + the duid slug for the unique id instead of the name.

regevbr replied:
indeed, fixed

Lash-L commented on tests/components/roborock/conftest.py:
```diff
@@ -67,19 +68,31 @@ async def update_values(
         return {prot: self.protocol_responses[prot] for prot in dyad_data_protocols}
 
 
+@pytest.fixture(name="bypass_api_client_fixture")
+def bypass_api_client_fixture() -> None:
+    """Skip calls to the API client."""
+    with (
+        patch(
+            "homeassistant.components.roborock.RoborockApiClient.get_home_data_v2",
+            return_value=HOME_DATA,
+        ),
+        patch(
+            "homeassistant.components.roborock.RoborockApiClient.get_scenes",
+            return_value=SCENES,
+        ),
+    ):
+        yield
```
Why'd you separate this out instead of just including it with the other fixture?

regevbr replied:
When I tried to add another patch, python errors - `SyntaxError: too many statically nested blocks` so I had to separate and nest

Lash-L commented on tests/components/roborock/test_scene.py:
```diff
@@ -0,0 +1,112 @@
+
+
+@pytest.fixture
+def bypass_api_client_get_scenes_fixture(bypass_api_fixture) -> None:
+    """Fixture to raise when getting scenes."""
+    with (
+        patch(
+            "homeassistant.components.roborock.RoborockApiClient.get_scenes",
+            side_effect=RoborockException(),
+        ),
+    ):
+        yield
```
This is only used once - i think it makes sense to just patch in the individual function

regevbr replied:
I couldn't patch it in the test body itself as it is need to be in the fixtures phase that creates the entities. In any case this test case is not relevant anymore so removed

regevbr replied:
This has become relevant again, so please let me know how do you think this can be improved

regevbr commented on homeassistant/components/roborock/button.py:
```diff
@@ -75,6 +101,16 @@ async def async_setup_entry(
+        *_build_setup_functions(config_entry.runtime_data.values()),
+        return_exceptions=True,
+    )
+    async_add_entities(
```
not relevant anymore

regevbr committed:
feature: upgrade python-roborock

regevbr committed:
feature: upgrade python-roborock

allenporter commented on homeassistant/components/roborock/coordinator.py:
```diff
@@ -168,6 +181,26 @@ async def set_current_map_rooms(self) -> None:
 
+    async def get_scenes(self) -> list[HomeDataScene]:
+        """Get scenes."""
+        return await self._api_client.get_scenes(
```
`execute_scene` catches exceptions and raises home assistant error but this does not. Can these be consistent?

regevbr replied:
sure

allenporter commented on homeassistant/components/roborock/scene.py:
```diff
@@ -0,0 +1,75 @@
+from .entity import RoborockEntity
+
+
+async def _setup_coordinator(
```
get_scenes is an async function so i don't think this is correct?

regevbr replied:
Not sure what you are referring to that is incorrect, if you mean the function name, I agree and changed it

allenporter replied:
I was referring tot he typing -- I may have misread where the `await` was, my mistake.

allenporter commented on homeassistant/components/roborock/scene.py:
```diff
@@ -0,0 +1,75 @@
+    coordinator: RoborockDataUpdateCoordinator,
+) -> list[SceneEntity]:
+    return [
+        RoborockSceneEntity(coordinator, scene)
```
I think this function should just return the scene data, and then create the entity in `async_setup_entry`. Name the fucntion `get_scenes` and have it return scenes. 

(The name `coordinator_entities` doesn't really mean anything, and this function `_setup_coordinator` has an ambiguous name given the coordinators are already setup)

regevbr replied:
I renamed the functions, that was an oversight on my end. Regarding the creation in the async setup entry - because the creations are coupled to both scene data and the coordinator, I'd rather keep it like today, but if you prefer to do it your way, I will need to make the get scene function return a tuple of coordinator and scene data, which I think is less clean. Let me know what you think

allenporter commented on homeassistant/components/roborock/scene.py:
```diff
@@ -0,0 +1,75 @@
+    async_add_entities(
+        entity
+        for entities in coordinator_entities
+        if isinstance(entities, list)
```
We before discussed the principles behind exception handling and not silently ignoring exceptions, and this is again silently ignoring exceptions.

regevbr replied:
handled

allenporter commented on homeassistant/components/roborock/scene.py:
```diff
@@ -0,0 +1,75 @@
+        )
+        self._scene_id = scene.id
+        self._coordinator = coordinator
+        self.entity_description = EntityDescription(
```
Consider passing this the `EntityDescription` the constructor argument since that is a pretty common pattern. (and if you swap out `_scene_id`  with `entity_description.key` then you don't need to accept the scene.

regevbr replied:
handled though there is a type mismatch so had to make a conversion

regevbr commented on homeassistant/components/roborock/coordinator.py:
```diff
@@ -168,6 +181,26 @@ async def set_current_map_rooms(self) -> None:
+    async def get_scenes(self) -> list[HomeDataScene]:
+        """Get scenes."""
+        return await self._api_client.get_scenes(
+            self._user_data, self.roborock_device_info.device.duid
```
my bad, fixed that and some other places that did that

regevbr committed:
feature: upgrade python-roborock

regevbr committed:
feature: upgrade python-roborock

regevbr committed:
feature: upgrade python-roborock

regevbr commented:
Hi guys, please let me know in your convenient time if there is anything else that needs fixing/changed

allenporter commented on homeassistant/components/roborock/scene.py:
```diff
@@ -0,0 +1,74 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up scene platform."""
+    devices_scene_entities = await asyncio.gather(
+        *[
+            _get_device_scenes(coordinator)
+            for coordinator in config_entry.runtime_data.v1
+        ],
+    )
+    async_add_entities(
+        entity
+        for device_scene_entities in devices_scene_entities
+        for entity in device_scene_entities
+    )
```
Something seems off to me creating the entity in another function. I think its better to create entities in `async_setup_entry` and closer to `async_add_entities`. 

I see the reason why you separated this is because the coordinator has multiple scenes, so I would suggest doing it this way to separate the scene fetching from the entity creation:

```suggestion
    scene_lists = await asyncio.gather(
        *(
            coordinator.get_scenes()
            for coordinator in config_entry.runtime_data.v1
        ),
    )

    async_add_entities(
        RoborockSceneEntity(
            coordinator,
            EntityDescription(
                key=str(scene.id),
                name=scene.name,
            ),
        )
        for coordinator, scenes in zip(config_entry.runtime_data.v1, scene_lists)
        for scene in scenes
    )
```

I may have some of the syntax off, but its something like that.

regevbr replied:
done

allenporter reviewed (changes requested):
Looks fine to me just a style request...

regevbr committed:
feature: upgrade python-roborock

allenporter commented on homeassistant/components/roborock/scene.py:
```diff
@@ -0,0 +1,64 @@
+            ),
+        )
+        for coordinator, scenes in zip(
+            config_entry.runtime_data.v1, scene_lists, strict=False
```
My understanding is that `strict=False` implies its ok for the lists to not have the same length. That should not happen. You added this which wasn't in the example i had -- can you remove or help me understand why this could possibly be OK? It would just introduce bugs if it were true, no?

regevbr replied:
This was added by automatically by Ruff and I failed to notice it. I chnaged it to True

allenporter reviewed (commented):
The force pushing, or whateve,r is making it difficult to review changes between review rounds.

regevbr committed:
feature: upgrade python-roborock

regevbr commented:
> The force pushing, or whateve,r is making it difficult to review changes between review rounds.

Yes Im aware, that is why I do a rebase so commit history is kept and should not affect the "changes since last review" methodology.

In any case I need to rebase again now cause there are conflicts with recent changes to his integration that were already emerged. How do you want me to proceed with it?

Lash-L commented:
Personally @regevbr - I just hit the resolve conflicts button in github or 'update branch' button in github.

regevbr commented:
> Personally @regevbr - I just hit the resolve conflicts button in github or 'update branch' button in github.

That is what i did

allenporter reviewed (changes requested):
Oh sorry, i see there there are still conflicts.

regevbr committed:
Merge branch 'dev' into roborock

regevbr committed:
feature: upgrade python-roborock

regevbr commented:
> Oh sorry, i see there there are still conflicts.

resolved

allenporter commented:
@regevbr we decided your original proposal to use buttons was correct.  However, given this was late in the release cycle we needed to revert so we can target 2025.4 with this. Are you open to sending another PR to add back to button? If you're annoyed at me understandable :) and i can do it but also wanted your name to be on the feature.
